### PR TITLE
Add opaque ports to destination service updates

### PR DESCRIPTION
**Pull request by kleimkuhler**
## Summary

This changes the destination service to start indicating whether a profile is an
opaque protocol or not.

Currently, profiles returned by the destination service are built by chaining
together updates coming from watching Profile and Traffic Split updates.

With this change, we now also watch updates to Opaque Port annotations on pods
and namespaces; if an update occurs this is now included in building a profile
update and is sent to the client.

## Details

Watching updates to Profiles and Traffic Splits is straightforward--we watch
those resources and if an update occurs on one associated to a service we care
about then the update is passed through.

For Opaque Ports this is a little different because it is an annotation on pods
or namespaces. To account for this, we watch the endpoints that we should care
about.

### When host is a Pod IP

When getting the profile for a Pod IP, we check for the opaque ports annotation
on the pod and the pod's namespace. If one is found, we'll indicate if the
profile is an opaque protocol if the requested port is in the annotation.

We do not subscribe for updates to this pod IP. The only update we really care
about is if the pod is deleted and this is already handled by the proxy.

### When host is a Service

When getting the profile for a Service, we subscribe for updates to the
endpoints of that service. For any ports set in the opaque ports annotation on
any of the pods, we check if the requested port is present.

Since the endpoints for a service can be added and removed, we do subscribe for
updates to the endpoints of the service.

Signed-off-by: Kevin Leimkuhler <kevin@kleimkuhler.com>

### CODE REVIEW

File: controller/api/destination/opaque_ports_adaptor.go [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+		merged = *opa.profile
+	}
+	merged.Spec.OpaquePorts = opa.opaquePorts
+	fmt.Printf("publishing SP: %v", merged)
+	opa.listener.Update(&merged)
+}
+
+func getOpaquePortsAnnotations(k8sAPI *k8s.API, pod *corev1.Pod) (map[uint32]bool, error) {
+	opaquePorts := make(map[uint32]bool)
+	obj, err := k8sAPI.GetObjects("", pkgk8s.Namespace, pod.Namespace, labels.Everything())
```
Comment (kleimkuhler) [inline controller/api/destination/opaque_ports_adaptor.go:95]: Checking for the opaque ports annotation on the namespace here is why the destination RBAC change is needed.

**Review by adleong — COMMENTED** [review]
Adding further watching logic to the destination controller always makes me nervous.  However, upon reading this closely, it's not nearly as bad as I feared.  The adapter pattern works really well here and updates propagate through in a predictable way.  I think what really saves us here is that all necessary watches can established when the request is received and we never have to do "chained" watches where we would subscribe or unsubscribe dynamically based on updates from other watches.

A few comments below, but very nice work.  :100:

File: controller/api/destination/opaque_ports_adaptor.go [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/labels"
+)
+
+type opaquePortsAdaptor struct {
+	listener    watcher.ProfileUpdateListener
+	k8sAPI      *k8s.API
+	log         *logging.Entry
+	profile     *sp.ServiceProfile
+	opaquePorts map[uint32]bool
```
[the remark this answers is not in the document]
Comment (kleimkuhler) [reply controller/api/destination/opaque_ports_adaptor.go:23]: Yep makes sense.

File: controller/api/destination/opaque_ports_adaptor.go [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+}
+
+func newOpaquePortsAdaptor(listener watcher.ProfileUpdateListener, k8sAPI *k8s.API, log *logging.Entry, opaquePorts map[uint32]bool) *opaquePortsAdaptor {
```
[the remark this answers is not in the document]
Comment (kleimkuhler) [reply controller/api/destination/opaque_ports_adaptor.go:26]: Ah right good catch.

File: go.sum [hunk 4]
```diff
@@ -77,8 +77,10 @@ github.com/VividCortex/gohistogram v1.0.0/go.mod h1:Pf5mBqqDxYaXu3hDrrU+w6nw50o/
 github.com/afex/hystrix-go v0.0.0-20180502004556-fa1af6a1f4f5/go.mod h1:SkGFH1ia65gfNATL8TAiHDNxPzPdmEL5uirI2Uyuz6c=
 github.com/agnivade/levenshtein v1.0.1/go.mod h1:CURSv5d9Uaml+FovSIICkLbAUZ9S4RqaHDIsdSBg7lM=
 github.com/alecthomas/template v0.0.0-20160405071501-a0175ee3bccc/go.mod h1:LOuyumcjzFXgccqObfd/Ljyb9UuFJ6TxHnclSeseNhc=
+github.com/alecthomas/template v0.0.0-20190718012654-fb15b899a751 h1:JYp7IbQjafoB+tBA3gMyHYHrpOtNuDiK/uB5uXxq5wM=
```
[the remark this answers is not in the document]
Comment (kleimkuhler) [reply go.sum:80]: These changes get added back when building the binary so I don't think they are left over from older changes in these commits.

**Review by adleong — APPROVED** [review]
🥛 🐇

File: controller/api/destination/server_test.go [hunk 5]
```diff
@@ -520,6 +537,90 @@ func TestGetProfiles(t *testing.T) {
...
+			t.Fatalf("Expected 'namespace' metric label to exist but it did not")
+		}
+		if first.Endpoint.ProtocolHint == nil {
+			t.Fatalf("Expected protocol hint but found none")
+		}
+		if !first.OpaqueProtocol {
+			t.Fatalf("Expected protocol to be opaque but it was not")
+		}
+		if first.Endpoint.Addr.String() != epAddr.String() {
+			// t.Fatalf("Expected endpoint IP to be %s, but it was %s", epAddr.Ip, first.Endpoint.Addr.Ip)
```
Comment (olix0r) [inline controller/api/destination/server_test.go:620]: rm?
```suggestion
```

**Comment by kleimkuhler** [comment]
I've added an integration test where where we expect to find the following metric:

```
tcp_open_total{peer=\"src\",direction=\"inbound\",tls=\"no_identity\",no_tls_reason=\"port_skipped\"} 1
```

This metric tells us that there is 1 open inbound TCP connection that is TLS'd because the port is skipped.

This metric occurs when the a workload's port (`8080` in this example) is marked as opaque and requests are sent to it from another injected workload (slow-cooker in this example).

**Review by alpeb — APPROVED** [review]
Sorry for the delayed review :grimacing: 
Very nice indeed, it all makes great sense to me. Thanks for the tests as well :heart:

**Pull request merged**